gremlin-go/driver/bytecode_test.go

```diff
@@ -0,0 +1,120 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+*/
+
+package gremlingo
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestBytecode(t *testing.T) {
+	t.Run("Constructor", func(t *testing.T) {
+		bc1 := newBytecode(nil)
+		assert.NotNil(t, bc1.bindings)
+		assert.NotNil(t, bc1.sourceInstructions)
+		assert.NotNil(t, bc1.stepInstructions)
+		assert.Empty(t, bc1.bindings)
+		assert.Empty(t, bc1.sourceInstructions)
+		assert.Empty(t, bc1.stepInstructions)
+
+		sourceInstructions := []instruction{{
+			operator:  "mockSource",
+			arguments: nil,
+		}}
+		stepInstructions := []instruction{{
+			operator:  "mockStep",
+			arguments: nil,
+		}}
+		bindingMap := make(map[string]interface{})
+		bindingMap["mock"] = 123
+		bc1.sourceInstructions = sourceInstructions
+		bc1.stepInstructions = stepInstructions
+		bc1.bindings = bindingMap
+
+		bc2 := newBytecode(bc1)
+		assert.NotNil(t, bc2.bindings)
+		assert.NotNil(t, bc2.sourceInstructions)
+		assert.NotNil(t, bc2.stepInstructions)
+		assert.Empty(t, bc2.bindings)
+		assert.Equal(t, sourceInstructions, bc2.sourceInstructions)
+		assert.Equal(t, stepInstructions, bc2.stepInstructions)
+	})
+
+	t.Run("addSource", func(t *testing.T) {
+		expectedSourceInstructions := []instruction{{
+			operator:  "mockSource",
+			arguments: []interface{}{123},
+		}}
+		bc := newBytecode(nil)
+		err := bc.addSource("mockSource", 123)
+		assert.Nil(t, err)
+		assert.Equal(t, expectedSourceInstructions, bc.sourceInstructions)
+	})
+
+	t.Run("addStep", func(t *testing.T) {
+		expectedStepInstructions := []instruction{{
+			operator:  "mockStep",
+			arguments: []interface{}{123},
+		}}
+		bc := newBytecode(nil)
+		err := bc.addStep("mockStep", 123)
+		assert.Nil(t, err)
+		assert.Equal(t, expectedStepInstructions, bc.stepInstructions)
```

Comment:
Shall we improve best practices around testing by adding messages to asserts? If tests get broken, it is easier to find the root cause from the test report if additional context is provided. Example from: https://pkg.go.dev/github.com/stretchr/testify/assert#Equal
`assert.Equal(t, a, b, "The two words should be the same.")`

Reply:
I don't think it is necessary here. The current failure specifies what test fails already, and a Equal check is rather straightforward on what it's checking for, as well as having a clear message of the failure (expected things to be equal but they are not).